File: result.go
package main

import (
	"bytes"
	"strconv"
)

//result contains an analysis of a set of commit
type result struct {
	//The name of the files seen
	files []string
	//The name of the files seen
	fileExtensions []string
	//How many region we have (i.e. seperated by @@)
	regions int
	//How many line were added total
	lineAdded int
	//How many line were deleted totla
	lineDeleted int
	//How many times the functionj seen in the code are called before and after
	functionCalls map[string]struct{ before, after int }
}

//String returns the value of results as a formated string
func (r *result) String() string {

	var buffer bytes.Buffer
	buffer.WriteString("Files: \n")
	for _, file := range r.files {
		buffer.WriteString("	-")
		buffer.WriteString(file)
		buffer.WriteString("\n")
	}
	buffer.WriteString("Extensions: \n")
	for _, ext := range r.fileExtensions {
		buffer.WriteString("\t-")
		buffer.WriteString(ext)
		buffer.WriteString("\n")
	}
	r.appendIntValueToBuffer(r.regions, "Regions", &buffer)
	r.appendIntValueToBuffer(r.lineAdded, "LA", &buffer)
	r.appendIntValueToBuffer(r.lineDeleted, "LD", &buffer)

	buffer.WriteString("Function calls (before, after): \n")
	for key, value := range r.functionCalls {
		buffer.WriteString("\t")
		buffer.WriteString(key)
		buffer.WriteString(" : ")
		buffer.WriteString(strconv.Itoa(value.before))
		buffer.WriteString(", ")
		buffer.WriteString(strconv.Itoa(value.after))
		buffer.WriteString("\n")
	}

	return buffer.String()
}

//appendIntValueToBuffer appends int value to a bytes buffer
func (r result) appendIntValueToBuffer(value int, label string, buffer *bytes.Buffer) {
	buffer.WriteString(label)
	buffer.WriteString(" : ")
	buffer.WriteString(strconv.Itoa(value))
	buffer.WriteString("\n")
}

File: src/result.rs
use std::fmt;
use std::collections::{HashSet, HashMap};
use std::ops::AddAssign;

// result contains an analysis of a set of commits
pub struct Result {
    // The name of the files seen
    files: HashSet<String>,
    // How many regions we have (i.e. seperated by @@)
    regions: usize,
    // How many lines were added total
    lineAdded: usize,
    // How many lines were deleted total
    lineDeleted: usize,
    // How many times the function seen in the code are called
    functionCalls: HashMap<String, usize>,
}

impl Result {
    // Creates an empty Result struct.
    pub fn empty() -> Self {
        Self::new(HashSet::new(), 0, 0, 0, HashMap::new())
    }

    // Constructor for a Result struct that expects everything to be handed to it.
    pub fn new(files: HashSet<String>,
               regions: usize,
               lineAdded: usize,
               lineDeleted: usize,
               functionCalls: HashMap<String, usize>)
               -> Self {
        Result {
            files,
            regions,
            lineAdded,
            lineDeleted,
            functionCalls,
        }
    }

    pub fn add_filename(&mut self, filename: String) {
        self.files.insert(filename);
    }

    pub fn add_function_call(&mut self, function: String) {
        let current = *self.functionCalls.get(&function).unwrap_or(&0);
        self.functionCalls.insert(function,current + 1);
    }

    pub fn add_region(&mut self) {
        self.regions += 1
    }

    pub fn count_added_line(&mut self) {
        self.lineAdded += 1
    }

    pub fn count_removed_line(&mut self) {
        self.lineDeleted += 1
    }
}

// Implementation of the formating system in rust ensuring that the structure may be
// printed in a way close to that of the print statment provided in the go structure.
impl fmt::Display for Result {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Files: \n")?;
        for file in &self.files {
            write!(f, "    -{}\n", file)?;
        }
        write!(f, "Regions: {}\n", self.regions)?;
        write!(f, "LA: {}\n", self.lineAdded)?;
        write!(f, "LD: {}\n", self.lineDeleted)?;

        write!(f, "Functions calls: \n")?;
        for (key, value) in &self.functionCalls {
            write!(f, "   {}: {}\n", key, value)?;
        }
        Ok(())
    }
}

// Iplementation of the += opperator which is useful syntacticaly to clearly demonstrait
// how results are combined.
impl AddAssign for Result {
    fn add_assign(&mut self, mut other: Result) {
        other
            .files
            .drain()
            .for_each(|file| { self.files.insert(file); });
        self.regions += other.regions;
        self.lineAdded += other.lineAdded;
        self.lineDeleted += other.lineDeleted;
        other
            .functionCalls
            .drain()
            .for_each(|(key, value)| {
                          let to_add = *self.functionCalls.get(&key).unwrap_or(&0);
                          self.functionCalls.insert(key, value + to_add);
                      });
    }
}
